Bluetooth: ATT: Multiple fixes

Vudentz opened the pull request:
ATT channel sent callback shall not be overwritting until the
operation completes as it can result in breaking flow control when
CONFIG_BT_ATT_ENFORCE_FLOW is enabled.

Fixes #25964
Fixes #26070
Fixes #26071

Signed-off-by: Luiz Augusto von Dentz <luiz.von.dentz@intel.com>

Vudentz committed:
Bluetooth: ATT: Fix overwritting sent callback

ATT channel sent callback shall not be overwritting until the
operation completes as it can result in breaking flow control when
CONFIG_BT_ATT_ENFORCE_FLOW is enabled.

Fixes #25964
Fixes #26071

Signed-off-by: Luiz Augusto von Dentz <luiz.von.dentz@intel.com>

cvinayak commented:
@alexsven could you check if your unsubscribe scenario under ATT traffic works now with this PR?

joerchan reviewed (dismissed):
This appears to fix a problem reported by some, but not the ATT timeout that was reported by @cvinayak 

I have created a new bug-report for this issue and updated the fixes line.

carlescufi reviewed (changes requested):
@Vudentz this PR fixes #26071 according to @joerchan, but it doesn't fix #25964. Can you update the commit to reflect this or add an additional commit to fix #25964 please?

Vudentz committed:
Bluetooth: ATT: Fix not processing pending requests

Since the TX semaphore is used for all types of PDUs a request may have
to be put on the request list while there is no pending request pending
which means no response will be generated to trigger att_process,
previously this condition was handled by setting the request as
currently pending and append its buffer to tx_queue but this is no
longer efficient since there could be more than one channel active the
code should try all of them before queueing back to request list.

To fix this the request list will now be processed each time a PDU has
been sent.

Fixes #26070

Signed-off-by: Luiz Augusto von Dentz <luiz.von.dentz@intel.com>

carlescufi reviewed (dismissed):
@Vudentz thanks!

>Fixes #25964 #26071

You will need 2 separate `Fixes` lines for GitHub to pick up both issues

carlescufi commented on subsys/bluetooth/host/att.c:
```diff
@@ -194,6 +202,8 @@ static void bt_att_sent(struct bt_l2cap_chan *ch)
 		chan->sent(chan);
 	}
```
Agreed, this looks suspicious, is `chan->sent` being set to NULL elsewhere?

Vudentz replied:
That is just there because chan_cb can return NULL, reseting to NULL doesn't make any difference here since chan_send will always attempt to set the callback when called.

Vudentz commented:
> @Vudentz thanks!
> 
> > Fixes #25964 #26071
> 
> You will need 2 separate `Fixes` lines for GitHub to pick up both issues

Done.

joerchan commented:
@Vudentz There should be 3 issues fixed now, right? Missing 26063

joerchan commented:
@Vudentz I also ran into an issue where the error code was not forwarded to the user because of this:
https://github.com/zephyrproject-rtos/zephyr/blob/master/subsys/bluetooth/host/att.c#L2833
I'm not sure why the error code was ignored in this case, are there some other parts relying on the fact that it returns 0 here?

Vudentz committed:
Bluetooth: ATT: Fix not returning error

bt_l2cap_send_cb may fail if there are no context available which means
that the request would not be sent, also due to the use of custom
callback it cannot be queued either so the only option is to return the
error and let the application handle it.

Signed-off-by: Luiz Augusto von Dentz <luiz.von.dentz@intel.com>

joerchan commented:
```
D: bt_att_recv():Received ATT chan 0x20007e28 code 0x07 len 4
D: att_handle_find_type_rsp():
D: att_handle_rsp():chan 0x20007e28 err 0x00 len 4: 0e001000
D: att_req_destroy():req 0x20007cb8
D: bt_att_req_free():req 0x20007cb8
D: gatt_find_type_rsp():err 0x00
D: gatt_find_type_rsp():start_handle 0x000e end_handle 0x0010
D: gatt_find_info():start_handle 0x000f end_handle 0x0010
D: bt_att_req_alloc():req 0x20007cb8
D: bt_att_req_send():conn 0x20002e08 req 0x20007cb8
D: bt_att_chan_req_send():req 0x20007cb8
W: chan_req_send():chan 0x20007e28 req 0x20007cb8 len 5
W: chan_req_send():231: buf 0x200080a0 ref
D: chan_send():code 0x04
W: chan_req_send():req->buf = NULL
D: bt_att_req_send():req 0x20007cb8 queued
D: att_sent():conn 0x20002e08 chan 0x20007e2c
D: bt_att_sent():chan 0x20007e28
D: chan_rsp_sent():chan 0x20007e28
W: chan_req_send():chan 0x20007e28 req 0x20007cb8 len 0
W: chan_req_send():231: buf 0x00000000 ref
ASSERTION FAIL [buf] @ WEST_TOPDIR/zephyr/subsys/net/buf.c:574

```
We end up with request in the queue that has set req->buf = NULL;
Compare the two functions below and their handling of the returned err value.
```
static int chan_req_send(struct bt_att_chan *chan, struct bt_att_req *req)
{
        /* ... */
	err = chan_send(chan, net_buf_ref(req->buf), NULL);
	if (err < 0) {
		net_buf_unref(req->buf);
		BT_WARN("req->buf = NULL");
		req->buf = NULL;
		chan->req = NULL;
	}

	return err;
}

```

```
int bt_att_req_send(struct bt_conn *conn, struct bt_att_req *req)
{
	/* ... */
	SYS_SLIST_FOR_EACH_CONTAINER_SAFE(&att->chans, chan, tmp, node) {
			/* ... */
			ret = bt_att_chan_req_send(chan, req);
			if (ret >= 0) {
				return ret;
			}
		}
	}

	/* Queue the request to be send later */
	sys_slist_append(&att->reqs, &req->node);

	BT_DBG("req %p queued", req);

	return 0;
}
```

Vudentz commented:
> We end up with request in the queue that has set req->buf = NULL;
> Compare the two functions below and their handling of the returned err value.
> 
> ```
> static int chan_req_send(struct bt_att_chan *chan, struct bt_att_req *req)
> {
>         /* ... */
> 	err = chan_send(chan, net_buf_ref(req->buf), NULL);
> 	if (err < 0) {
> 		net_buf_unref(req->buf);
> 		BT_WARN("req->buf = NULL");
> 		req->buf = NULL;

Ive remove the above statement, it shouldn't be reseting the buffer to NULL.

jhedberg commented on subsys/bluetooth/host/att.c:
```diff
@@ -316,6 +384,18 @@ static int bt_att_chan_send(struct bt_att_chan *chan, struct net_buf *buf,
+	err = bt_att_chan_send(chan, buf, cb);
+	if (err) {
+		/* Responses need to be sent back using the same channel */
+		k_fifo_put(&chan->tx_queue, buf);
```
This needs to be `net_buf_puf()`, so that fragment handling is correct. Even if you don't have fragments it ensures that buf->flags, etc is correct.

Vudentz replied:
Done.

Vudentz committed:
Bluetooth: ATT: Fix low throughput

ATT_PENDING_SENT does severely impact the throughput since multiple
packets no longer can be scheduled at same time, so instead of always
setting it regardless of the bearer/channel it is now only used for
EATT since that cannot set its own callbacks.

Signed-off-by: Luiz Augusto von Dentz <luiz.von.dentz@intel.com>

Vudentz committed:
Bluetooth: ATT: Fix using of k_fifo_{put,get}

These functions don't work with buffers that do have fragments, instead
this replaces their usage with net_buf_{put,get}.

Signed-off-by: Luiz Augusto von Dentz <luiz.von.dentz@intel.com>

Vudentz commented on subsys/bluetooth/host/att.c:
```diff
@@ -2420,6 +2485,11 @@ static void att_chan_detach(struct bt_att_chan *chan)
 	}
 
+	/* Release pending buffers */
+	while ((buf = k_fifo_get(&chan->tx_queue, K_NO_WAIT))) {
```
Done.

Vudentz committed:
Bluetooth: shell: Add support for printing the write rate

This makes the gatt metrics also available for
gatt write-without-rsp-cb so it now prints the rate of each write:

uart:~$ gatt write-without-response-cb 1e ff 10 10
Write #1: 16 bytes (0 bps)
Write #2: 32 bytes (3445948416 bps)
Write #3: 48 bytes (2596929536 bps)
Write #4: 64 bytes (6400 bps)
Write #5: 80 bytes (8533 bps)
Write #6: 96 bytes (10666 bps)
Write #7: 112 bytes (8533 bps)
Write #8: 128 bytes (9955 bps)
Write #9: 144 bytes (11377 bps)
Write #10: 160 bytes (7680 bps)
Write #11: 176 bytes (8533 bps)
Write #12: 192 bytes (9386 bps)
Write Complete (err 0)
Write #13: 208 bytes (8533 bps)
Write #14: 224 bytes (9244 bps)
Write #15: 240 bytes (9955 bps)
Write #16: 256 bytes (8000 bps)

Signed-off-by: Luiz Augusto von Dentz <luiz.von.dentz@intel.com>

joerchan commented on subsys/bluetooth/host/att.c:
```diff
@@ -178,36 +188,97 @@ static int chan_send(struct bt_att_chan *chan, struct net_buf *buf,
+	struct net_buf *buf;
+	int err;
+
+	while ((buf = net_buf_get(queue, K_NO_WAIT))) {
```
`while` with always one iteration? change to `if`?

carlescufi replied:
Why only one iteration? This will loop until `buf != NULL`? which btw, why `K_NO_WAIT` in this case and busy-loop like that?

Vudentz replied:
That indeed can be converted to if statement, I will change that.

Vudentz replied:
Done.

joerchan commented on subsys/bluetooth/host/att.c:
```diff
@@ -2420,6 +2537,11 @@ static void att_chan_detach(struct bt_att_chan *chan)
 		k_sem_give(&chan->tx_sem);
 	}
 
+	/* Release pending buffers */
```
In the ATT timeout handler the channel that the ATT timeout happened for will be detached, what about if you had other  ATT channels active at the same time? How would they be released?

Vudentz replied:
Timeout is per bearer, other channels will not be affected, provided they don't timeout as well.

joerchan replied:
But you would call att_reset and notify gatt that the connection is disconnect, not the channel. Since att_reset is called the att->conn is set to NULL and bt_att_send will now return ENOTCONN for everything as far as I can tell.

Vudentz replied:
Well read it again:

```
	/* Don't notify if there are still channels to be used */
	if (!sys_slist_is_empty(&att->chans)) {
		return;
	}
```

joerchan replied:
Right. missed that.

Vudentz commented:
@carlescufi process_queue will not attempt to take tx_sem like bt_att_chan_send would, so it is essencially just taking ownership of the slot used by the previous request, iirc this was made like that to reduce the time we spend on the sent callback.

carlescufi reviewed (approved):
Tested and verified to solve the issues linked.

carlescufi commented:
@Vudentz would you mind sending this PR against the `v2.3-branch` so we backport it in time for v2.3.1? Thanks!
